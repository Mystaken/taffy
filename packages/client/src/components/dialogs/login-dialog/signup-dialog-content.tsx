import React, { FunctionComponent } from 'react';
import {
  DialogContent,
  Grid,
  TextField,
  DialogActions,
  Button
} from '@material-ui/core';
import { useForm } from 'react-hook-form';
import { validateEmail } from '../../../utils/validators';
import { SignupDialogContentProps, FullSignupDetails } from './types';

export const SignupDialogContent: FunctionComponent<SignupDialogContentProps> = ({
  onSignup,
  disabled = false
}) => {
  const { register, errors, handleSubmit, watch } = useForm();

  const handleOnSubmit = (data: Record<string, any>) => {
    const formData = data as FullSignupDetails;
    onSignup?.({
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      password: formData.password
    });
  };

  return (
    <form onSubmit={handleSubmit(handleOnSubmit)}>
      <DialogContent>
        <Grid container direction="column">
          <TextField
            label="First Name"
            type="text"
            name="firstName"
            inputRef={register({ required: true, minLength: 1 })}
            error={!!errors.firstName}
            helperText={errors.firstName ? 'Required' : ''}
            disabled={disabled}
            autoFocus
          />
          <TextField
            label="Last Name"
            type="text"
            name="lastName"
            inputRef={register({ required: true, minLength: 1 })}
            error={!!errors.lastName}
            helperText={errors.lastName ? 'Required' : ''}
            disabled={disabled}
          />
          <TextField
            label="Email"
            type="text"
            name="email"
            inputRef={register({
              required: true,
              minLength: 3,
              validate: validateEmail
            })}
            error={!!errors.email}
            helperText={errors.email ? 'Invalid Email' : ''}
            disabled={disabled}
          />
          <TextField
            label="Password"
            type="password"
            name="password"
            inputRef={register({ required: true, minLength: 8 })}
            error={!!errors.password}
            helperText={
              errors.password ? 'Password must have at least 8 characters' : ''
            }
            disabled={disabled}
          />
          <TextField
            label="Confirm Password"
            type="password"
            name="confirmPassword"
            inputRef={register({
              required: true,
              validate: value => value === watch('password')
            })}
            error={!!errors.confirmPassword}
            helperText={errors.confirmPassword ? 'Password must match' : ''}
            disabled={disabled}
          />
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button color="primary" type="submit" disabled={disabled}>
          Sign Up
        </Button>
      </DialogActions>
    </form>
  );
};
