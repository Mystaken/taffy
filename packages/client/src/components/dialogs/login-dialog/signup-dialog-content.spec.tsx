import React from 'react';
import { create } from 'react-test-renderer';
import { SignupDialogContent } from './signup-dialog-content';
import { shallow } from 'enzyme';

jest.mock('react-hook-form', () => ({
  useForm: jest.fn(() => ({
    register: jest.fn(),
    errors: {},
    handleSubmit: jest.fn((d: any) => d),
    watch: jest.fn()
  }))
}));

describe('<SignupDialogContent />', () => {
  it('has rendered correctly', () => {
    const component = create(<SignupDialogContent />);
    const tree = component.toJSON()!;
    expect(tree).toMatchSnapshot();
  });
  it('should invoke onSubmit when submit button is clicked.', () => {
    const submitFn = jest.fn();
    const wrapper = shallow(<SignupDialogContent onSignup={submitFn} />);
    wrapper
      .find('form')
      .props()
      .onSubmit?.('' as any);
    expect(submitFn).toBeCalled();
  });
  it('should not invoke onSubmit when submit button is clicked with errors.', () => {
    const submitFn = jest.fn();
    const wrapper = shallow(<SignupDialogContent onSignup={submitFn} />);
    wrapper
      .find('form')
      .props()
      .onSubmit?.('' as any);
    expect(submitFn).toBeCalled();
  });
});
