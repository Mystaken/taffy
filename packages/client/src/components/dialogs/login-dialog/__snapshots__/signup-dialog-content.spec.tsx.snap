// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupDialogContent /> has rendered correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          First Name
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="firstName"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Last Name
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="lastName"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Email
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
        
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Confirm Password
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="confirmPassword"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
          />
        </div>
        
      </div>
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Sign Up
      </span>
    </button>
  </div>
</form>
`;
